sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_authentication_policy.py

```diff
@@ -0,0 +1,91 @@
+# coding=utf-8
+# ------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------
+from typing import TYPE_CHECKING
+
+from azure.core.exceptions import ServiceRequestError
+from azure.core.pipeline.policies import AsyncHTTPPolicy
+
+from ._async_exchange_client import ACRExchangeClient
+
+if TYPE_CHECKING:
+    from azure.core.credentials import TokenCredential
+    from azure.core.pipeline import PipelineRequest, PipelineResponse
+    from typing import Optional, Dict, Any, Union
+
+
+def _enforce_https(request):
+    # type: (PipelineRequest) -> None
+    """Raise ServiceRequestError if the request URL is non-HTTPS and the sender did not specify enforce_https=False"""
+
+    # move 'enforce_https' from options to context so it persists
+    # across retries but isn't passed to a transport implementation
+    option = request.context.options.pop("enforce_https", None)
+
+    # True is the default setting; we needn't preserve an explicit opt in to the default behavior
+    if option is False:
+        request.context["enforce_https"] = option
+
+    enforce_https = request.context.get("enforce_https", True)
+    if enforce_https and not request.http_request.url.lower().startswith("https"):
+        raise ServiceRequestError(
+            "Bearer token authentication is not permitted for non-TLS protected (non-https) URLs."
+        )
+
+
+class ContainerRegistryChallengePolicy(AsyncHTTPPolicy):
+    """Authentication policy for ACR which accepts a challenge"""
+
+    def __init__(self, credential, endpoint):
+        # type: (TokenCredential, str) -> None
+        super(ContainerRegistryChallengePolicy, self).__init__()
```

Comment:
Definitely Python 3 here
```suggestion
    def __init__(self, credential: "AsyncTokenCredential", endpoint: str) -> None:
        super().__init__()
```